File: native/android/CameraMeterPlugin.kt
package com.minilight.app

import android.content.Context
import android.graphics.ImageFormat
import android.hardware.camera2.CameraCaptureSession
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CaptureRequest
import android.hardware.camera2.CaptureResult
import android.hardware.camera2.TotalCaptureResult
import android.media.ImageReader
import android.os.Handler
import android.os.HandlerThread
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Real-metering plugin: runs a low-res Camera2 session and streams the
 * camera's own auto-exposure metadata plus the metered region's mean
 * luminance, so Dart can compute an absolute EV with no calibration.
 *
 * Copied verbatim into the generated android project by CI.
 */
class CameraMeterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var context: Context
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var sink: EventChannel.EventSink? = null

    private var thread: HandlerThread? = null
    private var handler: Handler? = null
    private var device: CameraDevice? = null
    private var session: CameraCaptureSession? = null
    private var reader: ImageReader? = null

    @Volatile private var regionMode = 1
    @Volatile private var pointNx: Double? = null
    @Volatile private var pointNy: Double? = null
    @Volatile private var lastExposureNs = 0L
    @Volatile private var lastIso = 100
    private var apertureF = 1.8f

    override fun onAttachedToEngine(b: FlutterPlugin.FlutterPluginBinding) {
        context = b.applicationContext
        methodChannel = MethodChannel(b.binaryMessenger, "minilight/exposure")
        methodChannel!!.setMethodCallHandler(this)
        eventChannel = EventChannel(b.binaryMessenger, "minilight/exposure_events")
        eventChannel!!.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(b: FlutterPlugin.FlutterPluginBinding) {
        stop()
        methodChannel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAvailable" -> result.success(true)
            "start" -> {
                regionMode = call.argument<Int>("regionMode") ?: 1
                start(); result.success(null)
            }
            "setRegion" -> {
                regionMode = call.argument<Int>("regionMode") ?: 1
                result.success(null)
            }
            "setPoint" -> {
                pointNx = call.argument<Double>("nx")
                pointNy = call.argument<Double>("ny")
                result.success(null)
            }
            "stop" -> { stop(); result.success(null) }
            else -> result.notImplemented()
        }
    }

    override fun onListen(args: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(args: Any?) { sink = null }

    private fun start() {
        if (device != null) return
        thread = HandlerThread("minilight-cam").also { it.start() }
        handler = Handler(thread!!.looper)
        val mgr = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
        val id = mgr.cameraIdList.firstOrNull {
            mgr.getCameraCharacteristics(it)
                .get(CameraCharacteristics.LENS_FACING) ==
                CameraCharacteristics.LENS_FACING_BACK
        } ?: mgr.cameraIdList.first()
        val chars = mgr.getCameraCharacteristics(id)
        chars.get(CameraCharacteristics.LENS_INFO_AVAILABLE_APERTURES)
            ?.firstOrNull()?.let { apertureF = it }

        reader = ImageReader.newInstance(320, 240, ImageFormat.YUV_420_888, 2)
        reader!!.setOnImageAvailableListener({ r ->
            val img = r.acquireLatestImage() ?: return@setOnImageAvailableListener
            try {
                val luma = meanLuma(img)
                emit(luma)
            } finally {
                img.close()
            }
        }, handler)

        try {
            mgr.openCamera(id, object : CameraDevice.StateCallback() {
                override fun onOpened(cam: CameraDevice) {
                    device = cam
                    val req = cam.createCaptureRequest(
                        CameraDevice.TEMPLATE_PREVIEW
                    ).apply {
                        addTarget(reader!!.surface)
                        set(
                            CaptureRequest.CONTROL_AE_MODE,
                            CaptureRequest.CONTROL_AE_MODE_ON
                        )
                    }
                    cam.createCaptureSession(
                        listOf(reader!!.surface),
                        object : CameraCaptureSession.StateCallback() {
                            override fun onConfigured(s: CameraCaptureSession) {
                                session = s
                                s.setRepeatingRequest(
                                    req.build(), captureCallback, handler
                                )
                            }
                            override fun onConfigureFailed(
                                s: CameraCaptureSession
                            ) {}
                        },
                        handler
                    )
                }
                override fun onDisconnected(cam: CameraDevice) { cam.close() }
                override fun onError(cam: CameraDevice, e: Int) { cam.close() }
            }, handler)
        } catch (e: SecurityException) {
            sink?.error("permission", "Camera permission denied", null)
        }
    }

    private val captureCallback = object : CameraCaptureSession.CaptureCallback() {
        override fun onCaptureCompleted(
            s: CameraCaptureSession,
            req: CaptureRequest,
            result: TotalCaptureResult
        ) {
            result.get(CaptureResult.SENSOR_EXPOSURE_TIME)?.let {
                lastExposureNs = it
            }
            result.get(CaptureResult.SENSOR_SENSITIVITY)?.let { lastIso = it }
            result.get(CaptureResult.LENS_APERTURE)?.let { apertureF = it }
        }
    }

    private fun meanLuma(img: android.media.Image): Double {
        val y = img.planes[0]
        val buf = y.buffer
        val rowStride = y.rowStride
        val w = img.width
        val h = img.height
        var left = 0; var top = 0; var rw = w; var rh = h
        val px = pointNx; val py = pointNy
        if (px != null && py != null) {
            val s = (minOf(w, h) * 0.08).toInt().coerceAtLeast(2)
            left = ((px * w).toInt() - s / 2).coerceIn(0, w - s)
            top = ((py * h).toInt() - s / 2).coerceIn(0, h - s)
            rw = s; rh = s
        } else if (regionMode == 0) {
            val s = (minOf(w, h) * 0.08).toInt().coerceAtLeast(2)
            left = (w - s) / 2; top = (h - s) / 2; rw = s; rh = s
        }
        var sum = 0.0; var count = 0
        var yy = top
        while (yy < top + rh) {
            var xx = left
            val base = yy * rowStride
            while (xx < left + rw) {
                sum += (buf.get(base + xx).toInt() and 0xFF)
                count++
                xx += 4
            }
            yy += 4
        }
        val mean8 = if (count == 0) 0.0 else sum / count / 255.0
        // De-gamma (approx sRGB) to a linear fraction.
        return Math.pow(mean8, 2.2)
    }

    private fun emit(linearLuma: Double) {
        val payload = HashMap<String, Any>()
        payload["exposureNs"] = lastExposureNs
        payload["iso"] = lastIso
        payload["apertureF"] = apertureF.toDouble()
        payload["midLuma"] = linearLuma
        Handler(context.mainLooper).post { sink?.success(payload) }
    }

    private fun stop() {
        try { session?.stopRepeating() } catch (_: Exception) {}
        session?.close(); session = null
        device?.close(); device = null
        reader?.close(); reader = null
        thread?.quitSafely(); thread = null; handler = null
    }
}

File: native/ios/CameraMeterPlugin.swift
import AVFoundation
import Flutter
import UIKit

/// Real-metering plugin: an AVCaptureSession streams the camera's own
/// auto-exposure metadata (exposure duration, ISO, aperture) plus the
/// metered region's mean luminance so Dart can derive an absolute EV with
/// no calibration. Copied into the generated Runner target by CI.
public class CameraMeterPlugin: NSObject, FlutterPlugin,
  FlutterStreamHandler, AVCaptureVideoDataOutputSampleBufferDelegate {

  private var sink: FlutterEventSink?
  private let session = AVCaptureSession()
  private var device: AVCaptureDevice?
  private let queue = DispatchQueue(label: "minilight.cam")
  private var regionMode = 1
  private var pointNx: Double?
  private var pointNy: Double?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = CameraMeterPlugin()
    let m = FlutterMethodChannel(
      name: "minilight/exposure",
      binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: m)
    let e = FlutterEventChannel(
      name: "minilight/exposure_events",
      binaryMessenger: registrar.messenger())
    e.setStreamHandler(instance)
  }

  public func handle(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    switch call.method {
    case "isAvailable":
      result(true)
    case "start":
      if let a = call.arguments as? [String: Any],
        let r = a["regionMode"] as? Int { regionMode = r }
      start()
      result(nil)
    case "setRegion":
      if let a = call.arguments as? [String: Any],
        let r = a["regionMode"] as? Int { regionMode = r }
      result(nil)
    case "setPoint":
      let a = call.arguments as? [String: Any]
      pointNx = a?["nx"] as? Double
      pointNy = a?["ny"] as? Double
      result(nil)
    case "stop":
      stop()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    sink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    sink = nil
    return nil
  }

  private func start() {
    queue.async {
      guard self.device == nil else { return }
      self.session.beginConfiguration()
      self.session.sessionPreset = .low
      guard
        let dev = AVCaptureDevice.default(
          .builtInWideAngleCamera, for: .video, position: .back),
        let input = try? AVCaptureDeviceInput(device: dev)
      else { return }
      self.device = dev
      if self.session.canAddInput(input) { self.session.addInput(input) }
      let output = AVCaptureVideoDataOutput()
      output.alwaysDiscardsLateVideoFrames = true
      output.videoSettings = [
        kCVPixelBufferPixelFormatTypeKey as String:
          kCVPixelFormatType_420YpCbCr8BiPlanarFullRange
      ]
      output.setSampleBufferDelegate(self, queue: self.queue)
      if self.session.canAddOutput(output) { self.session.addOutput(output) }
      self.session.commitConfiguration()
      self.session.startRunning()
    }
  }

  private func stop() {
    queue.async {
      if self.session.isRunning { self.session.stopRunning() }
      for i in self.session.inputs { self.session.removeInput(i) }
      for o in self.session.outputs { self.session.removeOutput(o) }
      self.device = nil
    }
  }

  public func captureOutput(
    _ output: AVCaptureOutput,
    didOutput sampleBuffer: CMSampleBuffer,
    from connection: AVCaptureConnection
  ) {
    guard let pb = CMSampleBufferGetImageBuffer(sampleBuffer),
      let dev = device else { return }
    CVPixelBufferLockBaseAddress(pb, .readOnly)
    defer { CVPixelBufferUnlockBaseAddress(pb, .readOnly) }
    let w = CVPixelBufferGetWidthOfPlane(pb, 0)
    let h = CVPixelBufferGetHeightOfPlane(pb, 0)
    guard let base = CVPixelBufferGetBaseAddressOfPlane(pb, 0) else { return }
    let stride = CVPixelBufferGetBytesPerRowOfPlane(pb, 0)
    let ptr = base.assumingMemoryBound(to: UInt8.self)

    var left = 0, top = 0, rw = w, rh = h
    if let px = pointNx, let py = pointNy {
      let s = max(2, Int(Double(min(w, h)) * 0.08))
      left = min(max(0, Int(px * Double(w)) - s / 2), w - s)
      top = min(max(0, Int(py * Double(h)) - s / 2), h - s)
      rw = s; rh = s
    } else if regionMode == 0 {
      let s = max(2, Int(Double(min(w, h)) * 0.08))
      left = (w - s) / 2; top = (h - s) / 2; rw = s; rh = s
    }

    var sum = 0.0, count = 0
    var y = top
    while y < top + rh {
      var x = left
      let row = y * stride
      while x < left + rw {
        sum += Double(ptr[row + x]); count += 1; x += 4
      }
      y += 4
    }
    let mean = count == 0 ? 0.0 : sum / Double(count) / 255.0
    let linear = pow(mean, 2.2)
    let expSeconds = CMTimeGetSeconds(dev.exposureDuration)
    let payload: [String: Any] = [
      "exposureNs": expSeconds * 1e9,
      "iso": Double(dev.iso),
      "apertureF": Double(dev.lensAperture),
      "midLuma": linear,
    ]
    DispatchQueue.main.async { self.sink?(payload) }
  }
}
